File: src/main/java/wang/crick/ddl2plantuml/convert/DdlExtractor.java
package wang.crick.ddl2plantuml.convert;

import com.alibaba.druid.sql.ast.SQLDataType;
import com.alibaba.druid.sql.ast.SQLExpr;
import com.alibaba.druid.sql.ast.statement.SQLColumnDefinition;
import com.alibaba.druid.sql.ast.statement.SQLSelectOrderByItem;
import com.alibaba.druid.sql.ast.statement.SQLTableElement;
import com.alibaba.druid.sql.dialect.mysql.ast.MySqlKey;
import com.alibaba.druid.sql.dialect.mysql.ast.MySqlPrimaryKey;
import com.alibaba.druid.sql.dialect.mysql.ast.MySqlUnique;
import com.alibaba.druid.sql.dialect.mysql.ast.statement.MySqlCreateTableStatement;
import com.alibaba.druid.sql.dialect.mysql.parser.MySqlStatementParser;
import wang.crick.ddl2plantuml.dictionary.IndexTypeEnum;
import wang.crick.ddl2plantuml.model.Column;
import wang.crick.ddl2plantuml.model.Index;
import wang.crick.ddl2plantuml.util.SqlStringUtil;
import wang.crick.ddl2plantuml.model.Table;
import wang.crick.ddl2plantuml.util.ExprUtil;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.toMap;

/**
 * ddl解析
 *
 * @author wangyuheng
 * @date 2018/6/28 11:13
 */
public class DdlExtractor {

    public Table extract(String sql) {
        MySqlStatementParser mySqlStatementParser = new MySqlStatementParser(sql);
        MySqlCreateTableStatement ddl = (MySqlCreateTableStatement) mySqlStatementParser.parseCreateTable();
        List<SQLTableElement> sqlTableElementList = ddl.getTableElementList();
        Table table = new Table();
        table.setName(SqlStringUtil.removeBacktick(ddl.getName().getSimpleName()));
        table.setComment(SqlStringUtil.removeSingleQuote(ExprUtil.getNullableString(ddl.getComment())));

        List<Column> columnList = sqlTableElementList.stream()
                .filter(element -> element instanceof SQLColumnDefinition)
                .map(element -> (SQLColumnDefinition) element)
                .map(scd -> {

                    String name = SqlStringUtil.removeBacktick(scd.getName().getSimpleName()).trim();
                    String comment = SqlStringUtil.removeSingleQuote(ExprUtil.getNullableString(scd.getComment())).trim();
                    //默认值
                    String defaultValue = Optional.ofNullable(scd.getDefaultExpr()).map(Object::toString).orElse(null);
                    //数据类型
                    SQLDataType dataType = scd.getDataType();
                    //类型大小信息
                    List<SQLExpr> dataTypeArguments = dataType.getArguments();
                    Integer size = null;
                    if (null != dataTypeArguments && dataTypeArguments.size() > 0) {
                        size = Integer.parseInt(dataTypeArguments.get(0).toString());
                    }
                    return new Column(name, comment, dataType.getName(), size, defaultValue, scd.containsNotNullConstaint());
                }).collect(Collectors.toList());
        table.setColumnList(columnList);

        Map<String, Column> nameColumnMap = columnList.stream().collect(toMap(ele -> ele.getName().toLowerCase(), Function.identity()));
        List<Index> indexList = sqlTableElementList.stream()
                .filter(element -> element instanceof MySqlKey)
                .map(element -> (MySqlKey) element)
                .map(key -> {
                    Index index;
                    if (key instanceof MySqlPrimaryKey) {
                        index = new Index(IndexTypeEnum.primary, ExprUtil.getNullableString(key.getComment()), getColumnListByKey(key, nameColumnMap));
                    } else if (key instanceof MySqlUnique) {
                        index = new Index(IndexTypeEnum.unique, ExprUtil.getNullableString(key.getComment()), getColumnListByKey(key, nameColumnMap));
                    } else {
                        index = new Index(IndexTypeEnum.ordinary, ExprUtil.getNullableString(key.getComment()), getColumnListByKey(key, nameColumnMap));
                    }
                    return index;
                }).collect(Collectors.toList());
        table.setIndexList(indexList);
        return table;
    }

    private String getLowerColumnName(SQLSelectOrderByItem item) {
        return SqlStringUtil.removeBacktick(item.getExpr().toString()).toLowerCase();
    }

    private List<Column> getColumnListByKey(MySqlKey key, Map<String, Column> nameColumnMap) {
        return key.getColumns().stream()
                .map(this::getLowerColumnName)
                .map(nameColumnMap::get)
                .collect(Collectors.toList());
    }

}

File: src/main/kotlin/wang/crick/ddl2plantuml/convert/DdlExtractor.kt
package wang.crick.ddl2plantuml.convert

import com.alibaba.druid.sql.ast.statement.SQLColumnDefinition
import com.alibaba.druid.sql.ast.statement.SQLTableElement
import com.alibaba.druid.sql.dialect.mysql.ast.MySqlKey
import com.alibaba.druid.sql.dialect.mysql.ast.MySqlPrimaryKey
import com.alibaba.druid.sql.dialect.mysql.ast.MySqlUnique
import com.alibaba.druid.sql.dialect.mysql.parser.MySqlStatementParser
import wang.crick.ddl2plantuml.dictionary.IndexTypeEnum
import wang.crick.ddl2plantuml.Column
import wang.crick.ddl2plantuml.Index
import wang.crick.ddl2plantuml.Table
import wang.crick.ddl2plantuml.util.SqlStringUtil
import java.util.*
import java.util.stream.Collectors

/**
 * ddl解析
 *
 * @author wangyuheng
 * @date 2019-01-20 12:53
 */
class DdlExtractor {

    companion object  {
        fun extract(sql: String): Table {
            val mySqlStatementParser = MySqlStatementParser(sql)
            val ddl = mySqlStatementParser.parseCreateTable()
            val sqlTableElementList = ddl.tableElementList

            val columnList = extractColumn(ddl.tableElementList)

            val nameColumnMap = columnList.map { it.name.toLowerCase() to it }.toMap()

            val indexList = sqlTableElementList.stream()
                    .filter { element -> element is MySqlKey }
                    .map { element -> element as MySqlKey }
                    .map { key ->
                        when (key) {
                            is MySqlPrimaryKey -> Index(IndexTypeEnum.PRIMARY, Objects.toString(key.comment, ""), getColumnListByKey(key, nameColumnMap))
                            is MySqlUnique -> Index(IndexTypeEnum.UNIQUE, Objects.toString(key.comment, ""), getColumnListByKey(key, nameColumnMap))
                            else -> Index(IndexTypeEnum.ORDINARY, Objects.toString(key.comment, ""), getColumnListByKey(key, nameColumnMap))
                        }
                    }.collect(Collectors.toList())

            return Table(
                    SqlStringUtil.removeBacktick(ddl.name.simpleName),
                    SqlStringUtil.removeSingleQuote(Objects.toString(ddl.comment, "")),
                    columnList,
                    indexList
            )
        }

        private fun extractColumn(tableElementList: List<SQLTableElement>): List<Column> {
            return tableElementList.stream()
                    .filter { element -> element is SQLColumnDefinition }
                    .map { element -> element as SQLColumnDefinition }
                    .map { scd ->
                        val name = SqlStringUtil.removeBacktick(scd.name.simpleName).trim()
                        val comment = SqlStringUtil.removeSingleQuote(Objects.toString(scd.comment, "")).trim()
                        //默认值
                        val defaultValue = scd.defaultExpr?.toString()
                        //数据类型
                        val dataType = scd.dataType
                        //类型大小信息
                        val dataTypeArguments = dataType.arguments
                        var size = 0
                        if (null != dataTypeArguments && dataTypeArguments.size > 0) {
                            size = Integer.parseInt(dataTypeArguments[0].toString())
                        }
                        Column(name, comment, dataType.name, size, defaultValue, scd.containsNotNullConstaint())
                    }.collect(Collectors.toList())
        }

        private fun getColumnListByKey(key: MySqlKey, nameColumnMap: Map<String, Column>): List<Column?> {
            return key.columns.stream()
                    .map { column -> SqlStringUtil.removeBacktick(column.expr.toString()).toLowerCase() }
                    .map { column -> nameColumnMap[column] }
                    .collect(Collectors.toList())
        }
    }




}